Refactor parts of the GameSpeed system

reaperrr opened the pull request:
Refactors how we read and apply game speeds to get rid of various hardcodings, work-arounds and limitations.

Fixes #19138.
Fixes hardcoded Timestep 40 (25 ticks/sec) assumption in `DateTimeGlobal`.
Removes various hardcodings and work-arounds.

Mailaender commented on OpenRA.Game/World.cs:
```diff
@@ -180,7 +180,19 @@ internal World(ModData modData, Map map, OrderManager orderManager, WorldType ty
+			{
+				var defaultGameSpeed = modData.Manifest.Get<GameSpeeds>().Speeds.FirstOrDefault(s => s.Key == "default");
+				if (defaultGameSpeed.Value != null)
+					Timestep = defaultGameSpeed.Value.Timestep;
```
I don't really like this, but it seems you have no choice as `MapOptions` is not available here. You are looking for a hard-coded key (maybe put it into constant somewhere) and silently fail when it is not found. At least add:

```csharp
				else
					Timestep = new GameSpeed().Timestep;
```

otherwise, timestep may stay 0 for mods that called it "Default" (which is the official backfall) or "normal".

pchote replied:
It looks like we should be able to do this more cleanly at the bottom of `MapOptions.Created`. This already knows what the mod-defined default speed is, so no need to hardcode "default".

It would also be good to elaborate a bit more about "at the right time", so I suggest the following:
```csharp
// The world timestep and order latency for regular games is hacked as a
// side-effect on top of the regular lobby options plumbing.
// The Shellmap and map editor bypass that plumbing, so we must manually fix
// the game speed here. The order latency is not relevant for these so is ignored.
if (self.World.Type != WorldType.Regular)
	self.World.Timestep = GameSpeed.Timestep;
```

Am i missing something that breaks if we allow the timestep to stay at the incorrect GlobalSettings default between world creation and `MapOptions.Created`?

Mailaender commented on OpenRA.Mods.Common/Scripting/Global/DateTimeGlobal.cs:
```diff
@@ -20,11 +21,15 @@ namespace OpenRA.Mods.Common.Scripting
 			tlm = context.World.WorldActor.TraitOrDefault<TimeLimitManager>();
+			var defaultGameSpeed = Game.ModData.Manifest.Get<GameSpeeds>().Speeds.FirstOrDefault(s => s.Key == "default");
+			var defaultTimestep = defaultGameSpeed.Value != null ? defaultGameSpeed.Value.Timestep : 40;
+			ticksPerSecond = 1000 / defaultTimestep;
```
Why don't you use
```csharp
context.World.WorldActor.Trait<MapOptions>().GameSpeed.Timestep
```
?

reaperrr replied:
That would change behavior of all time-based triggers!
`ticksPerSecond` must be identical between game speeds.

Mailaender commented on OpenRA.Mods.Common/Scripting/Global/DateTimeGlobal.cs:
```diff
@@ -20,11 +21,15 @@ namespace OpenRA.Mods.Common.Scripting
 		{
 			tlm = context.World.WorldActor.TraitOrDefault<TimeLimitManager>();
+			var defaultGameSpeed = Game.ModData.Manifest.Get<GameSpeeds>().Speeds.FirstOrDefault(s => s.Key == "default");
+			var defaultTimestep = defaultGameSpeed.Value != null ? defaultGameSpeed.Value.Timestep : 40;
```
Not sure if it is a good idea to hard-code `40` in here. Maybe use `new GameSpeed().Timestep`?

pchote replied:
Do we have a usecase for mods not defining a default gamespeed?
IMO we should just make this a requirement and not have the fallback at all.

pchote replied:
How about something like this (note: not tested ingame)?

This would avoid both hardcoding the "default" name and the fallback

```csharp
var gameSpeed = context.World.WorldActor.Info.TraitInfoOrDefault<MapOptionsInfo>()?.GameSpeed;
if (gameSpeed == null)
	throw new LuaException("Unable to determine the default game speed.");

if (!Game.ModData.Manifest.Get<GameSpeeds>().Speeds.TryGetValue(gameSpeed, out var info))
	throw new LuaException("Game speed '{0}' not defined.");

ticksPerSecond = 1000 / info.Timestep;
```

reaperrr replied:
I saw this too late for my update, but I'm not convinced it would be the right approach.

First of all, it would still kind of lead to https://github.com/OpenRA/OpenRA/pull/19297#discussion_r604356357

Furthermore, from what I've seen `MapOptions.GameSpeed` is only populated on actor creation, so fairly late.
At least in cases like shellmaps (possibly in other situations as well), that leads to `MapOptions.GameSpeed` not necessarily matching the actual gamespeed, so there's still a risk of a mismatch between what the modder expects vs. what the game actually does.
Using the default speed as base at least ensures that ticks per `DateTime.Second` are consistent across all gamespeeds.

pchote replied:
Right, makes sense.

We've been going around in circles trying to avoid the mess that I made when I first implemented the feature (sorry...). How about this idea to cut through some of the mess?

* Change the speed definitions in mod.yaml to look something like:
   ```
   GameSpeeds:
   	DefaultSpeed: default
   	Speeds:
   		slowest:
   			Name: Slowest
   			Timestep: 80
   			OrderLatency: 2
   		# etc...
   ```
   Change the `LoadSpeeds` method to something like this to parse it:
   ```csharp
   var speedNodes = y.Nodes.FirstOrDefault(n => n.Key == "Speeds");
   if (speedNodes == null)
   	throw new Exception("blah");

   foreach (var node in speedNodes.Value.Nodes)
   ```
* Change the definition in `MapOptions` to
   ```csharp
   [Desc("Default game speed (leave empty to use the default defined in mod.yaml).")]
   public readonly string GameSpeed = null;
   ```
* Remove `Timestep` and `OrderLatency` completely from the `Global` class. Instead define a `string GameSpeed = null` and change the hack in `LobbyCommands` to set this instead.
* Change the `World` ctor to set `World.Timestep` and a new `World.OrderLatency` based on `gameSpeeds.Speeds[orderManager.LobbyInfo.GlobalSettings.GameSpeed ?? gameSpeeds.DefaultSpeed]`
* Change the things that reference `GlobalSettings.Timestep` and `GlobalSettings.OrderLatency` to instead take them from `World`. This won't work for replays, but they should be able to fetch the values based on `GlobalSettings.GameSpeed` in the ctor instead. Lua should of course also do its own thing ignoring the `GlobalSettings` value.

This would hopefully centralize the game speed control so that the default specified in mod.yaml is always used unless the lobby specifies an override, also avoiding magic names.

reaperrr replied:
OK, combined with your other comment it makes sense.
It did work locally on the RA shellmap, so I'll update accordingly for now.

pchote replied:
Ugh, crossed comments again!

What do you think about my latest suggestion?

reaperrr replied:
That looks a lot better than what we have. I'll see if I run into any issues implementing that and go to IRC if I do to avoid more comment-crossing.

pchote replied:
Another quick thought to improve things further: ~~we might be able to introduce a new interface (`IGameSpeed`?) that the world ctor can use to query `MapOptions` (which is in Mods.Common, so not available to directly query) for its game speed~~. This would allow us to avoid `GlobalSettings.GameSpeed` completely and the serverside hack to set it

~~We'd then fall back to `gameSpeeds.DefaultSpeed` if there isn't exactly one trait implementing `IGameSpeed` or it returns a null value.~~

Note that i'm going to be afk/mobile for the next hour or two, so may not be able to respond on IRC immediately.

pchote replied:
A simpler solution that should give most of the same benefits will be to ignore/remove `GlobalSettings.GameSpeed` and the `LobbyCommands` hack as above, but instead of a new interface we just hardcode the "gamespeed" option in the world ctor with something like:
```csharp
var gameSpeedName = orderManager.LobbyInfo.GlobalSettings.OptionOrDefault("gamespeed", gameSpeeds.DefaultSpeed);
var gameSpeed = gameSpeeds.Speeds[gameSpeedName];
```

pchote commented:
The changelog entry when merging this should also mention #19298.

reaperrr committed:
Refactor GameSpeed setting

*Remove internal GameSpeed defaults
 Enforce setting values explicitly all the time
 Require definition of a DefaultSpeed

*Remove Global.Timestep default

*Remove the hacky Timestep/OrderLatency setting via LobbyInfo

*Fix shellmaps ignoring mod-defined gamespeeds

*Make DateTimeGlobal use the MapOptions gamespeed

reaperrr commented:
Updated with various improvements:

- Set `Session.Timestep` to something that makes it easier to notice when no gamespeed was loaded. I was thinking about removing the internal defaults completely, but didn't want to risk this crashing people's PCs or something if the game locks up on a Timestep 0. 
- Removed internal default values for `GameSpeed` and added requirement to set values explicitly all the time.
- Require mods to define a `default` gamespeed.
- Moved the shellmap override to `StartGame` as discussed on IRC before. My previous attempt only failed because I set the values too late (after World was populated) within that method.
- Adjusted `DateTimeGlobal` to the above changes.

Mailaender commented on OpenRA.Game/GameSpeed.cs:
```diff
@@ -28,8 +28,27 @@ public class GameSpeeds : IGlobalModData
+			if (defaultCount == 0)
+				throw new YamlException("Mod doesn't define a 'default:' gamespeed! (Note: Name can still be anything).");
+			else if (defaultCount > 1)
+				throw new YamlException("Mod defines too many 'default:' gamespeeds!");
```
I think this will already throw a key already exists in the dictionary exception.

pchote commented:
Switching labels to reflect that i'd like us to explore the best solution, but don't yet have a single clear set of fixups to apply (note: discussion is happening in the inline comment threads further up).

reaperrr commented:
~The last suggestions appear to work fine and I guess it's rare for players to set a custom `MapOptions.Gamespeed` on map rules, so we can probably ignore that edge case and leave the modder to figure out if the result is what they wanted.
Updated to apply suggestions for now.~
Edit: Too slow again...

reaperrr commented:
Updated.

Known issues (apart from `Global.OrderLatency` still existing): Gamespeed lobby option doesn't seem to be able to read the default speed anymore (not sure where we'd have to set that).

pchote commented on OpenRA.Game/GameSpeed.cs:
```diff
@@ -10,26 +10,44 @@
+
+			foreach (var node in speedsNode.Value.Nodes)
+			{
+				var gs = FieldLoader.Load<GameSpeed>(node.Value);
```
We can use the FieldLoader plumbing from trait loading to improve this error handling:
```csharp
try
{
	ret.Add(node.Key, FieldLoader.Load<GameSpeed>(node.Value));
}
catch (FieldLoader.MissingFieldsException e)
{
	var label = e.Missing.Length > 1 ? "Required properties missing" : "Required property missing";
	throw new YamlException("Error parsing GameSpeed {0}: {1}: {2}".F(node.Key, label, e.Missing.JoinWith(", ")));
}
```

then add `[FieldLoader.Require]` to the fields in GameSpeed.

reaperrr replied:
Done locally, resolving to improve overview.

pchote commented on OpenRA.Game/Network/Session.cs:
```diff
@@ -222,8 +222,11 @@ public class Global
+
+			// Only set for and used by replays now.
+			// TODO: Get rid of this entirely, if possible.
+			public int OrderLatency = 3;
```
Would adding this at the bottom of the `ReplayConnection` ctor work (I haven't tested it):
```csharp
var gameSpeeds = Game.ModData.Manifest.Get<GameSpeeds>();
var gameSpeedName = LobbyInfo.GlobalSettings.OptionOrDefault("gamespeed", gameSpeeds.DefaultSpeed);
orderLatency = gameSpeeds.Speeds[gameSpeedName].OrderLatency;
ordersFrame = orderLatency;
```

where `orderLatency` would be a new `readonly int` that can then also be used in `SendSync`.

pchote replied:
It looks like we can remove `UnitOrders.SetOrderLag` and instead change `Game.StartGame` to use

```csharp
using (new PerfTimer("NewWorld"))
{
	OrderManager.World = new World(ModData, map, OrderManager, type);
	OrderManager.FramesAhead = OrderManager.World.OrderLatency;
}
```

but again not tested - need to make sure that game saving and loading in particular doesn't break.

reaperrr replied:
Both suggestions seem to work fine locally (for savegames and replays, too), resolving to improve overview.

pchote commented on OpenRA.Game/World.cs:
```diff
@@ -180,7 +182,21 @@ internal World(ModData modData, Map map, OrderManager orderManager, WorldType ty
+			OriginalTimestep = Timestep = gameSpeed.Timestep;
+			OrderLatency = gameSpeed.OrderLatency;
+
+			// HACK: Turn down the latency if there is only one real player (or only bots)
```
`NonBotClients` also counts spectators, so I'm not sure this can ever be 0?

reaperrr replied:
Fixed locally, resolving to improve overview.

pchote commented on OpenRA.Game/World.cs:
```diff
@@ -36,6 +36,8 @@ public sealed class World : IDisposable
 
 		readonly Queue<Action<World>> frameEndActions = new Queue<Action<World>>();
 
+		public readonly int OriginalTimestep;
```
This shakes another skeleton out of the closet: manipulating the `Timestep` is a horrible hack that forces a bunch of places in the UI code to manually manage their own timestep variables to avoid incorrect timers - doing a "Find Usages" on `MapOptions.GameSpeed` will show these.

At a minimum it would be a good simple extra cleanup if we can remove these DIY workarounds and have these places use `world.OriginalTimestep` directly...

... but actually, it would be much better if we could instead have a single readonly `Timestep` and introduce a `float ReplayTimestepModifier` for replays. I *think* (but not 100% sure) that the only place where we actually care about this modifier is in `Game.InnerLogicTick`, which could instead do:

```csharp
var worldTimestep = world == null ? Timestep :
	world.IsLoadingGameSave ? 1 :
	world.IsReplay ? world.ReplayTimestepModifier * world.Timestep :
	world.Timestep;
```

Not tested and quite possibly regresses things, but would go a long way further to improving the timestep mess if it works!

reaperrr replied:
I've been struggling to mentally map the modifier to the other places it would affect (there's a few more, most notably `ReplayControlBarLogic`) plus how to avoid rounding down to zero unless explicitly wanted.
For me it was easier to just rename `OriginalTimestep` to `Timestep` and `Timestep` to `ReplayTimestep` and ensure the latter is exclusively used for replays. Hope that's good enough.

pchote commented on OpenRA.Mods.Common/Traits/World/MapOptions.cs:
```diff
@@ -123,10 +123,16 @@ void INotifyCreated.Created(Actor self)
 
-			var speed = self.World.LobbyInfo.GlobalSettings
-				.OptionOrDefault("gamespeed", info.GameSpeed);
+			var speeds = Game.ModData.Manifest.Get<GameSpeeds>();
```
This can simplify down to
```csharp
var gameSpeeds = Game.ModData.Manifest.Get<GameSpeeds>();
var gameSpeedName = self.World.LobbyInfo.GlobalSettings.OptionOrDefault("gamespeed", gameSpeeds.DefaultSpeed);
GameSpeed = gameSpeeds.Speeds[gameSpeedName];
```
which then matches the way we do it in the World ctor...

...but actually, I would go a step further by removing this from here and making the copy we set up in `World` public instead. We just need to change the comment
```csharp
// NOTE: The server hardcodes special-case logic for this option id
```
above to something like
```csharp
// NOTE: This is just exposing the UI, the backend logic for this option is hardcoded in World
```

Also note that all our uses of this will go away when cleaning up the UI timestep workarounds I mentioned above.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/Ingame/GameTimerLogic.cs:
```diff
@@ -37,15 +37,15 @@ public GameTimerLogic(Widget widget, OrderManager orderManager, World world)
 				var timestep = world.Timestep;
 				if (world.IsReplay)
-					timestep = world.WorldActor.Trait<MapOptions>().GameSpeed.Timestep;
+					timestep = world.OriginalTimestep;
```
This is one of the cases I was referring to above: we can use `world.OriginalTimestep` directly for both replays and non-replays below (or, even better, `world.Timestep` after fixing the replay speed implementation).

reaperrr replied:
This is pretty much solved regardless of whether we go for a replay speed modifier or replay-exclusive timestep, so resolving here to improve overview.

reaperrr commented:
Updated.

Known issue: Gamespeed lobby option still defaults to 'Not available'.
If someone can point me in the right direction before I trial & error my way through the UI code, I wouldn't mind.

pchote commented on OpenRA.Mods.Common/Traits/World/MapOptions.cs:
```diff
@@ -87,15 +87,15 @@ IEnumerable<LobbyOption> ILobbyOptions.LobbyOptions(Ruleset rules)
-			// NOTE: The server hardcodes special-case logic for this option id
+			// NOTE: This is just exposing the UI, the backend logic for this option is hardcoded in World
 			yield return new LobbyOption("gamespeed", GameSpeedDropdownLabel, GameSpeedDropdownDescription, GameSpeedDropdownVisible, GameSpeedDropdownDisplayOrder,
 				new ReadOnlyDictionary<string, string>(gameSpeeds), GameSpeed, GameSpeedDropdownLocked);
```
The "Not Available" happens because we pass the null default of `GameSpeed` in as the initial state.

Tweaking the variable definitions above and passing `GameSpeed ?? gameSpeed.DefaultGameSpeed` like in the other places will fix it.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -776,9 +784,16 @@ static void Loop()
-				var logicInterval = worldRenderer != null && worldRenderer.World.Timestep != 0 ? worldRenderer.World.Timestep : Ui.Timestep;
+				// Ideal time between logic updates.
+				var logicInterval = Ui.Timestep;
+				if (worldRenderer != null)
```
The inner `world == null` check can be removed by changing this to `if (worldRenderer?.world != null)`.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -776,9 +784,16 @@ static void Loop()
+					var world = worldRenderer.World;
+					logicInterval = world == null ? Ui.Timestep :
+						world.IsLoadingGameSave ? 1 :
+						world.IsReplay ? world.ReplayTimestep :
```
This should probably be `world.IsReplay && world.ReplayTimestep != 0` to avoid problems with paused replays.

pchote replied:
IMO it would be a good idea to keep the comment about `ReplayTimestep = 0 means the game is paused` too.

pchote commented on OpenRA.Mods.Common/Widgets/ObserverProductionIconsWidget.cs:
```diff
@@ -56,7 +56,7 @@ public ObserverProductionIconsWidget(World world, WorldRenderer worldRenderer)
 			this.worldRenderer = worldRenderer;
 			clocks = new Dictionary<ProductionQueue, Animation>();
-			timestep = world.IsReplay ? world.WorldActor.Trait<MapOptions>().GameSpeed.Timestep : world.Timestep;
+			timestep = world.Timestep;
```
Can we please remove `timestep` completely and use `world.Timestep` directly at the one place in `Draw()`?

pchote commented on OpenRA.Mods.Common/Widgets/ObserverSupportPowerIconsWidget.cs:
```diff
@@ -57,8 +57,6 @@ public ObserverSupportPowerIconsWidget(World world, WorldRenderer worldRenderer)
 
 			// Timers in replays should be synced to the effective game time, not the playback time.
 			timestep = world.Timestep;
```
Likewise here.

pchote commented on OpenRA.Mods.Common/Widgets/SupportPowerTimerWidget.cs:
```diff
@@ -40,8 +40,6 @@ public SupportPowerTimerWidget(World world)
 
 			// Timers in replays should be synced to the effective game time, not the playback time.
 			timestep = world.Timestep;
```
and here.

pchote reviewed (commented):
A couple more things I noticed on a first pass. This otherwise looks like i would expect, but need to start testing for regressions still.

The "Not Available" fix will also depend on rebasing this after the `ReadOnlyDictionary` removal just merged.

reaperrr commented:
Rebased and updated.

pchote commented on OpenRA.Game/Game.cs:
```diff
@@ -776,9 +784,17 @@ static void Loop()
+					var world = worldRenderer.World;
+					logicInterval = world.IsLoadingGameSave ? 1 :
+						world.IsReplay && world.ReplayTimestep != 0 ? world.ReplayTimestep :
+						world.IsReplay && world.ReplayTimestep == 0 ? Ui.Timestep :
```
Style nit: my previous suggestion here was explicitly aiming to avoid reassigning `UI.Timestep` to the variable again. I agree that this case does seem more correct to use `Ui.Timestep`, but it does mess up my original idea.

How about changing this to
```csharp
var logicInterval = Ui.Timestep;
var logicWorld = worldRenderer?.World;

// ReplayTimestep = 0 means the replay is paused: we need to keep logicInterval as UI.Timestep to avoid breakage
if (logicWorld != null && !(logicWorld.IsReplay && logicWorld.ReplayTimestep == 0))
	logicInterval = logicWorld.IsLoadingGameSave ? 1 :
		logicWorld.IsReplay ? logicWorld.ReplayTimestep :
		logicWorld.Timestep;
```
instead?

pchote reviewed (dismissed):
Code looks good (except for one small style nit below), and I couldn't find any obvious problems when testing skirmish/bots/saving/replays/multiplayer.

reaperrr commented:
Fixed style nit.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/ef4f362731f89dbb33a4924ddf471d37577dc8db
